kroxylicious/src/main/java/io/kroxylicious/proxy/internal/ResponseOrderer.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.internal;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+
+import io.kroxylicious.proxy.frame.Frame;
+
+/**
+ * While processing a request from a Client, we want to enable custom Protocol
+ * Filters to decide to send a response toward the Client instead of forwarding
+ * the message on towards the upstream broker. In this case we may not be able
+ * to immediately send the response to the Client. There may be other requests
+ * in-flight that have already been forwarded to the upstream broker and the
+ * Kafka protocol requires responses to be sent to the Client in the same order
+ * they sent the requests.
+ * This class tries to ensure that we respond to requests in the correct order by
+ * remembering which correlationIds are in-flight and enqueuing any responses we
+ * get that are out of order. Then, when we do encounter the response for the
+ * oldest in-flight correlationId we can check if there are any enqueued
+ * responses that can now be forwarded towards the client.
+ */
+public class ResponseOrderer extends ChannelDuplexHandler {
+
+    Deque<Integer> inflightCorrelationIds = new ArrayDeque<>();
```

Comment:
I'm wondering if these should be maintained as a channel property? Are there uses cases where Filters mights care if there are pending requests?

Reply:
I can't think of a use case

Reply:
yeah, I guess its quite contrived to think of a filter choosing to only take an async action if and only iff auto read is already off.